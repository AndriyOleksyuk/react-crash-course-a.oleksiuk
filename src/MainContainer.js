import React, { Component } from 'react';
import { Button, Card, Alert, Row, Col, Container } from 'react-bootstrap';

export class MainContainer extends Component {
    render() {
        const {text, cancelHandler, requestHandler, isRequestSended} = this.props;
        return (
            <Container>
                <Row className="show-grid" float="center">
                    <Col xs={12} xsOffset={6}>
                        <Card>
                            <Card.Header>Featured</Card.Header>
                            <Card.Body>
                                <Card.Title>First lesson</Card.Title>
                                <Button onClick={requestHandler} variant="primary" size="lg" block disabled={isRequestSended}>Get Email</Button>
                                <Button onClick={cancelHandler} variant="danger" size="lg" block disabled={!isRequestSended}>Cancel request</Button>
                            </Card.Body>
                            <Card.Body>
                                {text &&
                                    <Alert variant="success">
                                        Email: {text}
                                    </Alert>
                                }
                            </Card.Body>
                        </Card>
                    </Col>
                </Row>
            </Container>
        );
    }
}

export default MainContainer;
