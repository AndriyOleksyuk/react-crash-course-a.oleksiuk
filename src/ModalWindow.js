import React, { Component } from 'react';
import { Button, Modal } from 'react-bootstrap';

export class ModalWindow extends Component {
    render() {
        const {text, resendHandler, closeModalHandler} = this.props;
        return (
            <Modal show={!!text} onHide={closeModalHandler}>
                <Modal.Header>
                    <Modal.Title>Request failed</Modal.Title>
                </Modal.Header>

                <Modal.Body>
                    <p>{text}</p>
                </Modal.Body>

                <Modal.Footer>
                    <Button onClick={resendHandler} variant="secondary">Resend</Button>
                </Modal.Footer>
            </Modal>
        );
    }
}

export default ModalWindow;
