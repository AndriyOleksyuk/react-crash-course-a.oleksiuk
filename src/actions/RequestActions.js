import axios from 'axios';
import types from './Types';

export const cancelRequest = () => (dispatch, getState) => {
    let { source } = getState().request;
    const CancelToken = axios.CancelToken;

    source.cancel('Operation canceled by the user.');
   
    source = CancelToken.source();
    dispatch(setCancelSource(source));
}

export const fetchUserData = () => (dispatch, getState) => {
    let { source } = getState().request;
    const URL = 'https://randomuser.me/api/';
    const cancelToken = source.token;

    dispatch(startFetchUserData());

    return axios.get(URL, {cancelToken})
        .then(res => res.data)
        .then(function (data) {
            var userInfo = data.results[0];            
            dispatch(finishFetchUserData({email: userInfo.email, data: JSON.stringify(data, null, 4)}));
        }).catch(function (thrown) {
            if (axios.isCancel(thrown)) {
                dispatch(finishFetchUserData({}));
            } else {
                dispatch(handleError(thrown.message));
            }
        });
};

export const setCancelSource = (source) => ({ type: types.SET_CANCEL_TOKEN, source});
export const startFetchUserData = () => ({ type: types.START_FETCH_DATA});
export const finishFetchUserData = (response) => ({ type: types.FINISH_FETCH_DATA, response});
export const handleError = (error) => ({ type: types.HANDLE_ERROR, error});