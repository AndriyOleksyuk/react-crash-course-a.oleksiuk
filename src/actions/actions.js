import axios from 'axios';
import types from './types';

export const cancelRequest = (source) => (dispatch) => {
    source.cancel('Operation canceled by the user.');

    const CancelToken = axios.CancelToken;
    source = CancelToken.source();
    dispatch(setCancelSource(source));
}

export const fetchUserData = (source) => (dispatch) => {
    const URL = 'https://randomuser.me/api/';
    const cancelToken = source.token;

    dispatch(startFetchUserData());

    axios.get(URL, {cancelToken})
        .then(res => res.data)
        .then(function (data) {
            var userInfo = data.results[0];            
            dispatch(finishFetchUserData({email: userInfo.email, data: JSON.stringify(data, null, 4)}));
        }).catch(function (thrown) {
            if (axios.isCancel(thrown)) {
                dispatch(finishFetchUserData({}));
            } else {
                dispatch(handleError(thrown.message));
            }
        });
};

export const setCancelSource = (source) => ({ type: types.SET_CANCEL_TOKEN, source});
export const startFetchUserData = () => ({ type: types.START_FETCH_DATA});
export const finishFetchUserData = (response) => ({ type: types.FINISH_FETCH_DATA, response});
export const handleError = (error) => ({ type: types.HANDLE_ERROR, error});
export const setBtnSize = (size) => ({type: types.SET_BUTTONS_SIZE, size});
export const setBtnOutline = (outlineValue) => ({type: types.SET_BUTTONS_OUTLINE, outlineValue});