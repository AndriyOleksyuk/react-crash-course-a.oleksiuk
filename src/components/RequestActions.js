import React, {useContext, useRef, useEffect} from 'react';
import { Button} from 'react-bootstrap';
import {withSize} from '../components/Size';
import Outlined from '../components/Outlined'

const RequestActions = (props) => {
    const outlined = useContext(Outlined);
    const cancelBtn = useRef(null);
    const {requestHandler, size, isRequestSended, cancelHandler} = props;

    useEffect(() => {
        if(isRequestSended) {
            cancelBtn.current.focus();
        }
    }, [isRequestSended]);

    return (
        <>
            <Button onClick={requestHandler} variant={outlined + "primary"} size={size} block disabled={isRequestSended}>Get Email</Button>
            <Button onClick={cancelHandler} ref={cancelBtn} variant={outlined + "danger"} size={size} block disabled={!isRequestSended}>Cancel request</Button>
        </>
    );
}

export default withSize(RequestActions);
