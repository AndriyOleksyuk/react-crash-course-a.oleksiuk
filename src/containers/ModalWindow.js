import React from 'react';
import { Button, Modal } from 'react-bootstrap';

const ModalWindow = props => {
    const {text, resendHandler, closeModalHandler} = props;
    return (
        <Modal show={!!text} onHide={closeModalHandler}>
            <Modal.Header>
                <Modal.Title>Request failed</Modal.Title>
            </Modal.Header>

            <Modal.Body>
                <p>{text}</p>
            </Modal.Body>

            <Modal.Footer>
                <Button onClick={resendHandler} variant="secondary">Resend</Button>
            </Modal.Footer>
        </Modal>
    );
}

export default ModalWindow;
