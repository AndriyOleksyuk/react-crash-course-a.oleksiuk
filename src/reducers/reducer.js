export const reducer = (state, action) => {
    if (action.type === 'SET_BUTTONS_SIZE') {
        return {
            ...state,
            btnSize: action.size
        }
    } else if (action.type === 'SET_BUTTONS_OUTLINE') {
        return {
            ...state,
            btnOutline: action.outlineValue
        }
    } else if (action.type === 'START_FETCH_DATA') {
        return {
            ...state,
            requestIsSended: true,
            error: ''
        }
    } else if (action.type === 'FINISH_FETCH_DATA') {
        return {
            ...state,
            serviceData: action.response,
            requestIsSended: false
        }
    } else if (action.type === 'HANDLE_ERROR') {
        return {
            ...state,
            requestIsSended: false,
            error: action.error
        }
    } else if (action.type === 'SET_CANCEL_TOKEN') {
        return {
            ...state,
            source: action.source
        }
    }

    return state;
}