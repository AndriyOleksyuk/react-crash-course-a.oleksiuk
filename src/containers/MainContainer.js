import React, {Suspense, useState} from 'react';
import { Card, Row, Col, Container, ToggleButtonGroup, ToggleButton } from 'react-bootstrap';
import CardBody from './CardBody';

import {Size} from '../components/Size';
import Outlined from '../components/Outlined';

const MainContainer = props => {
    const {text, cancelHandler, requestHandler, isRequestSended} = props;
    const [btnSize, setBtnSize] = useState('sm');
    const [btnOutline, setBtnOutline] = useState('');

    const CardBodyAlert = React.lazy(() => import('../components/CardBodyAlert'));

    return (
        <Container>
            <Row className="show-grid" float="center">
                <Col xs={12} xsOffset={6}>
                    <Card>
                        <Card.Header>Featured</Card.Header>
                        <ToggleButtonGroup type="radio" defaultValue={btnSize} name="sizeSwitcher" onChange={setBtnSize}>
                            <ToggleButton variant="secondary" value={'sm'}>Set small buttons by HOC</ToggleButton>
                            <ToggleButton variant="secondary" value={'lg'}>Set big buttons by HOC</ToggleButton>
                        </ToggleButtonGroup>
                        <ToggleButtonGroup type="radio" defaultValue={btnOutline} name="outlinedSwitcher" onChange={setBtnOutline}>
                            <ToggleButton variant="outline-secondary" value={'outline-'}>set outline by hook useContext</ToggleButton>
                            <ToggleButton variant="outline-secondary" value={''}>delete outline by hook useContext</ToggleButton>
                        </ToggleButtonGroup>
                        
                        <Outlined.Provider value={btnOutline}>
                            <Size.Provider value={btnSize}>
                                <CardBody 
                                    text={text} 
                                    cancelHandler={cancelHandler} 
                                    requestHandler={requestHandler} 
                                    isRequestSended={isRequestSended} />
                            </Size.Provider>
                        </Outlined.Provider>

                        <Suspense fallback="">
                            <CardBodyAlert text={text} />
                        </Suspense>

                    </Card>
                </Col>
            </Row>
        </Container>
    );
}

export default MainContainer;
