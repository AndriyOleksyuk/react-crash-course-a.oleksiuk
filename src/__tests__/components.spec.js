import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import axios from 'axios';
import MainContainer from '../containers/MainContainer';
import ModalWindow from '../containers/ModalWindow';
import RequestActions from '../components/RequestActions';
import {fetchUserData} from '../actions/RequestActions';
import thunk from 'redux-thunk'
import { Provider } from 'react-redux';
import { createStore, applyMiddleware } from 'redux';
import reducer from '../reducers'
import initialState from '../state'

const store = createStore(
    reducer,
    initialState,
    applyMiddleware(thunk)
);


describe('<MainContainer />', () => {
    it('MainContainer should render and match snapshot', () => {
        const { container } = render(<Provider store={store}><MainContainer /></Provider>);

        expect(container).toMatchSnapshot();
    })
});


describe('<ModalWindow />', () => {
    it('ModalWindow should render and match snapshot', () => {
        const { container } = render(<Provider store={store}><ModalWindow /></Provider>);

        expect(container).toMatchSnapshot();
    })
});



describe('<RequestActions />', () => {
    it('RequestActions should render and match snapshot', () => {
        const { container } = render(<Provider store={store}><RequestActions /></Provider>);

        expect(container).toMatchSnapshot();
    })
});

describe('ModalWindow after success axios call', () => {

    jest.spyOn(axios, 'get').mockImplementationOnce(() => Promise.resolve({
        data: {
            results: [
                {
                    email: 'lucia.vandepavert@example.com'
                }
            ]
        }
    }));

    it('Should match snapshot', async () => {
        const {getByText, container } = render(<Provider store={store}><MainContainer /></Provider>);
        fireEvent.click(getByText('Get Email'));
        expect(axios.get).toHaveBeenCalledTimes(1);
        await Promise.resolve();
        expect(container).toMatchSnapshot();
    })
})

describe('ModalWindow after failed axios call', () => {

    it('Should match snapshot', async () => {
        jest.spyOn(axios, 'get').mockRejectedValue(new Error('Network error'));
        const {container } = render(<Provider store={store}><ModalWindow /></Provider>);

        await store.dispatch(fetchUserData());
        expect(container).toMatchSnapshot();
    })

})

