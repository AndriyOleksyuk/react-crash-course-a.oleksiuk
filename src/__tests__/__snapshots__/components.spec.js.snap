// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MainContainer /> MainContainer should render and match snapshot 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="show-grid row"
      float="center"
    >
      <div
        class="col-12"
        xsoffset="6"
      >
        <div
          class="card"
        >
          <div
            class="card-header"
          >
            Featured
          </div>
          <div
            class="btn-group btn-group-toggle"
            role="group"
          >
            <label
              class="btn active btn-secondary"
            >
              <input
                autocomplete="off"
                checked=""
                name="sizeSwitcher"
                type="radio"
                value="sm"
              />
              Set small buttons
            </label>
            <label
              class="btn btn-secondary"
            >
              <input
                autocomplete="off"
                name="sizeSwitcher"
                type="radio"
                value="lg"
              />
              Set big buttons
            </label>
          </div>
          <div
            class="btn-group btn-group-toggle"
            role="group"
          >
            <label
              class="btn btn-outline-secondary"
            >
              <input
                autocomplete="off"
                name="outlinedSwitcher"
                type="radio"
                value="outline-"
              />
              Set outline
            </label>
            <label
              class="btn active btn-outline-secondary"
            >
              <input
                autocomplete="off"
                checked=""
                name="outlinedSwitcher"
                type="radio"
                value=""
              />
              Delete outline
            </label>
          </div>
          <div
            class="card-body"
          >
            <div
              class="card-title h5"
            >
              First lesson
            </div>
            <button
              class="btn btn-primary btn-block btn-sm"
              type="button"
            >
              Get Email
            </button>
            <button
              class="btn btn-danger btn-block btn-sm"
              disabled=""
              type="button"
            >
              Cancel request
            </button>
          </div>
          
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<ModalWindow /> ModalWindow should render and match snapshot 1`] = `<div />`;

exports[`<RequestActions /> RequestActions should render and match snapshot 1`] = `
<div>
  <button
    class="btn btn-primary btn-block btn-sm"
    type="button"
  >
    Get Email
  </button>
  <button
    class="btn btn-danger btn-block btn-sm"
    disabled=""
    type="button"
  >
    Cancel request
  </button>
</div>
`;

exports[`ModalWindow after failed axios call Should match snapshot 1`] = `
<div
  aria-hidden="true"
/>
`;

exports[`ModalWindow after success axios call Should match snapshot 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="show-grid row"
      float="center"
    >
      <div
        class="col-12"
        xsoffset="6"
      >
        <div
          class="card"
        >
          <div
            class="card-header"
          >
            Featured
          </div>
          <div
            class="btn-group btn-group-toggle"
            role="group"
          >
            <label
              class="btn active btn-secondary"
            >
              <input
                autocomplete="off"
                checked=""
                name="sizeSwitcher"
                type="radio"
                value="sm"
              />
              Set small buttons
            </label>
            <label
              class="btn btn-secondary"
            >
              <input
                autocomplete="off"
                name="sizeSwitcher"
                type="radio"
                value="lg"
              />
              Set big buttons
            </label>
          </div>
          <div
            class="btn-group btn-group-toggle"
            role="group"
          >
            <label
              class="btn btn-outline-secondary"
            >
              <input
                autocomplete="off"
                name="outlinedSwitcher"
                type="radio"
                value="outline-"
              />
              Set outline
            </label>
            <label
              class="btn active btn-outline-secondary"
            >
              <input
                autocomplete="off"
                checked=""
                name="outlinedSwitcher"
                type="radio"
                value=""
              />
              Delete outline
            </label>
          </div>
          <div
            class="card-body"
          >
            <div
              class="card-title h5"
            >
              First lesson
            </div>
            <button
              class="btn btn-primary btn-block btn-sm"
              type="button"
            >
              Get Email
            </button>
            <button
              class="btn btn-danger btn-block btn-sm"
              disabled=""
              type="button"
            >
              Cancel request
            </button>
          </div>
          <div
            class="card-body"
          >
            <div
              class="fade alert alert-success show"
              role="alert"
            >
              Email: 
              lucia.vandepavert@example.com
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
